### PR TITLE
Refactoring project

### PR DESCRIPTION
**Pull request by lucasolsi-wex**

### CODE REVIEW

File: internal/validation/validate_user.go [hunk 1]
```diff
@@ -8,70 +8,69 @@ import (
...
+		cause := models.Causes{
 			Field:   "firstName/lastName",
 			Message: "User first/last names required",
 		}
 		errorCauses = append(errorCauses, cause)
-		return custom_errors.NewUserValidationFieldsError("Invalid fields", errorCauses)
+		return models.NewUserValidationFieldsError("Invalid fields", errorCauses)
 	}
 	return nil
 }
```
Comment (fabottega) [inline internal/validation/validate_user.go:63]: Validation logic related to the structure and constraints of the data can be placed within the model itself, ensuring that the data adheres to certain rules. You can have a `request.Validate()`, which should validate all the fields and return a list of validation errors.

File: internal/validation/validate_user.go [hunk 2]
```diff
@@ -8,70 +8,69 @@ import (
...
+		cause := models.Causes{
 			Field:   "firstName/lastName",
 			Message: "User with the same first and last name already exists",
 		}
 		errorCauses = append(errorCauses, cause)
-		return custom_errors.NewUserValidationFieldsError("Invalid fields", errorCauses)
+		return models.NewUserValidationFieldsError("Invalid fields", errorCauses)
 	}
 	return nil
 }
```
Comment (fabottega) [inline internal/validation/validate_user.go:76]: Services are responsible for encapsulating business logic. This includes processing requests, coordinating with repositories for data access, and performing validations related to the business rules. Any validation errors that are not strictly related to data integrity (handled by the models) can be managed within the service layer.

Comment (fabottega) [reply internal/validation/validate_user.go:76]: Your validation errors should adhere to the built in error interface. Instead of returning a custom error from the model, consider returning a standard error. You also can create custom validation error types for each one.

```go
func (ud *userDomainService) CreateUser(request models.UserRequest) (*models.UserResponse, error) {

        valErrors := request.Validate()
        if valErros ...
        ...
        // validate if user already exists
        
	userFromRepo, err := ud.repository.CreateUser(request)

	if err != nil {
		return nil, err
	}

	return userFromRepo, nil
}
```

File: internal/validation/validate_user.go (hunk 1 above)
Comment (fabottega) [reply internal/validation/validate_user.go:63]: Remember to validate all the fields.

File: internal/models/custom_errors.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package custom_errors
```
Comment (fabottega) [inline internal/models/custom_errors.go:1]: `func NewCustomErr(message, err string, code int, causes []Causes) *CustomErr`  isnt in use.

Comment (fabottega) [reply internal/models/custom_errors.go:1]: Why include the HTTP status code in your error? Is it intended to be serialized and sent back to the user? Consider the possibility of redundant information, as you've already set the HTTP status code on the request. It might be more appropriate for your controller to determine which error code to return to the user based on the errors returned by your service.

File: cmd/user-api/main.go [hunk 4]
```diff
@@ -4,11 +4,11 @@ import (
 	"context"
```
[the remark this answers is not in the document]
Comment (fabottega) [reply cmd/user-api/main.go:4]: Would also be nice to implement some kind of API versioning.

[the remark this answers is not in the document]
Comment (lucasolsi-wex) [reply cmd/user-api/main.go:4]: Done! Graceful shutdown implemented, handling database connection error, initializing and versioning API routes.

File: internal/models/custom_errors.go (hunk 3 above)
Comment (lucasolsi-wex) [reply internal/models/custom_errors.go:1]: You're right! I've removed that field and I'm not repeating the status code anymore.

File: internal/validation/validate_user.go [hunk 5]
```diff
@@ -8,70 +8,69 @@ import (
...
 func init() {
 	if val, ok := binding.Validator.Engine().(*validator.Validate); ok {
 		en := en_US.New()
 		un := ut.New(en, en)
-		error_translator, _ = un.GetTranslator("en")
-		en2.RegisterDefaultTranslations(val, error_translator)
+		errorTranslator, _ = un.GetTranslator("en")
+		en2.RegisterDefaultTranslations(val, errorTranslator)
 	}
 }
```
[the remark this answers is not in the document]
Comment (lucasolsi-wex) [reply internal/validation/validate_user.go]: It wasn't a requirement. Since it wasn't making that much of a difference, I've removed it.

File: internal/service/user_interface.go [hunk 6]
```diff
@@ -1,9 +1,8 @@
 package service
```
[the remark this answers is not in the document]
Comment (lucasolsi-wex) [reply internal/service/user_interface.go:1]: I've worked on it and I've moved the logic from other layers to the service layer.
All of the naming now is consistent. I've removed the unnecessary interface and refactored it as well. Unit tests are included now.

File: internal/repository/user_repository.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+			log.Fatal(err)
+		}
+		results = append(results, result)
+
+		if len(results) == 1 {
+			return true
+		}
+	}
+	return false
+}
```
Comment (fabottega) [inline internal/repository/user_repository.go:49]: This can be simplified. `cursor.Next(context.Background())` being true already indicates that a user exists.

```go
func (userRepo *userRepository) ExistsByFirstNameAndLastName(firstName, lastName string) bool {
// get collection
// create cursor with your query and defer Close
return cursor.Next(ctx)
}

File: internal/repository/user_repository.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+	return existingUser, err
+}
+
+func (userRepo *userRepository) CreateUser(request models.UserModel) (*models.UserModel, error) {
+	collection := userRepo.databaseConnection.Collection(MongoDBUserDb)
+
+	entity := models.NewUser(request.FirstName, request.LastName, request.Email, request.Age)
+
+	result, err := collection.InsertOne(context.Background(), entity)
```
Comment (fabottega) [inline internal/repository/user_repository.go:68]: Improve the use of context: Consider using the parent context instead of creating a new background context when making database queries. This allows the query to be canceled or timed out along with the parent context, ensuring proper resource cleanup and preventing potential context leaks.

File: internal/repository/user_repository.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	return false
+}
+
+func (userRepo *userRepository) FindUserById(id string) (*models.UserModel, error) {
+	collection := userRepo.databaseConnection.Collection(MongoDBUserDb)
+
+	existingUser := &models.UserModel{}
+
+	objectId, _ := primitive.ObjectIDFromHex(id)
```
Comment (fabottega) [inline internal/repository/user_repository.go:56]: We need proper error handling for converting the string ID to an ObjectID, and clarified the error message.

**Review by fabottega — COMMENTED** [review]
Please revisit the separation of concerns in your codebase. When implementing an architecture such as MVC or layered architecture, it's crucial to carefully consider how you are dividing and organizing your logic. Ensure that each component (models, views, controllers) has a clear and distinct responsibility, promoting maintainability and readability.

1. I see your validation logic spread across services and controller. Controllers should be lightweight and focus on managing the flow of data.
2. Models should define your business logic and the models. These should represent the entities in your application. I only see one entity, which is the user. Request and response are part of your service and are not models.
2.2. A good amount of the validation logic can be done inside the model. Your model can have a Validate receiver that validate all the user fields according to the business rules. Try to encapsulate as much business logic as possible withing the model.

File: internal/repository/user_repository.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	filter := bson.D{{Key: "_id", Value: objectId}}
+	err := collection.FindOne(context.Background(), filter).Decode(existingUser)
+
+	return existingUser, err
+}
+
+func (userRepo *userRepository) CreateUser(request models.UserModel) (*models.UserModel, error) {
+	collection := userRepo.databaseConnection.Collection(MongoDBUserDb)
+
+	entity := models.NewUser(request.FirstName, request.LastName, request.Email, request.Age)
```
Comment (fabottega) [inline internal/repository/user_repository.go:66]: Isn't the user passed in the func already a model.User? What is the point in converting it again?

Comment (lucasolsi-wex) [reply internal/repository/user_repository.go:66]: My bad, you're completely right. I was just testing you hahaha

File: internal/controller/user_controller.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+type userControllerInterface struct {
+	service service.UserInterfaceService
+}
+
+func (uc *userControllerInterface) CreateUser(gc *gin.Context) {
+	var userRequest models.UserRequest
+
+	if err := gc.ShouldBindJSON(&userRequest); err != nil {
+		log.Printf("Error trying to marshal object! Error=%s\n", err.Error())
```
Comment (fabottega) [inline internal/controller/user_controller.go:31]: In this part, log.Printf is used to log the error along with a custom message. You can replace it with log.Println if you prefer. The %v verb is used to format the error, and it will include the error type and message.

```go
if err != nil {
    log.Println("Error trying to marshal object: %v", err)
    ...
}
```

File: internal/models/user_model.go [hunk 12]
```diff
@@ -3,7 +3,7 @@ package models
 import "go.mongodb.org/mongo-driver/bson/primitive"
 
 type UserModel struct {
-	Id        primitive.ObjectID `json:"id,omitempty"`
+	Id        primitive.ObjectID `json:"id,omitempty" bson:"_id,omitempty"`
 	Email     string             `json:"email,omitempty"`
 	FirstName string             `json:"firstName,omitempty"`
 	LastName  string             `json:"lastName,omitempty"`
```
Comment (fabottega) [inline internal/models/user_model.go:9]: Does it make sense to mark these values with `omitempty` if we are required by business logic to populate most of them?

File: internal/service/user_interface.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/lucasolsi-wex/go-crud/internal/repository"
+	"github.com/lucasolsi-wex/go-crud/internal/validation"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"log"
+)
+
+type UserInterfaceService struct {
+	Repository repository.UserRepository
+}
```
Comment (fabottega) [inline internal/service/user_interface.go:17]: Isn't this the user service? Wouldn't it be better to name it "UserService"?

File: internal/service/user_interface.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+		return nil, models.NewInternalServerError(errorMessage)
+	}
+
+	return models.FromEntity(*existingUser), nil
+}
+
+func (us UserInterfaceService) CreateUser(request models.UserRequest, ctx *gin.Context) (*models.UserResponse, *models.CustomErr) {
+
+	if err := validation.ValidateFirstAndLastName(request); err != nil {
+		log.Print("Error while creating user. First and/or LastName is empty")
```
Comment (fabottega) [inline internal/service/user_interface.go:46]: Aim for consistent logging practices by centralizing log statements within the HTTP handler rather than scattering them across multiple locations. 
Here we have part of the errors being logged on the service side and the other part on the handler side.
This approach enhances maintainability and readability.

**Review by fabottega — COMMENTED** [review]
I recommend making a few minor tweaks, but overall, it seems good to me.

File: cmd/user-api/main.go [hunk 15]
```diff
@@ -0,0 +1,47 @@
+package main
```
Comment (fabottega) [inline cmd/user-api/main.go:1]: Please, update your docker files/Readme to match this change.

File: internal/models/user_model.go (hunk 12 above)
Comment (lucasolsi-wex) [reply internal/models/user_model.go:9]: You're right. I've removed that from all of the fields but id.

File: internal/service/user_interface.go (hunk 13 above)
Comment (lucasolsi-wex) [reply internal/service/user_interface.go:17]: Sure.

**Review by helios-matheuspires-wex — APPROVED** [review]
